pennylane/wires.py

```diff
@@ -14,48 +14,35 @@
 """
 This module contains the :class:`Wires` class, which takes care of wire bookkeeping.
 """
-from collections.abc import Sequence, Iterable
-import numpy as np
+from collections.abc import Iterable
 from numbers import Number
+import numpy as np
+
+from orderedset import OrderedSet
 
 
 class WireError(Exception):
     """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires."""
 
 
 def _process(wires):
-    """Converts the input to a tuple of numbers or strings."""
+    """ Generates a tuple from whatever nested structure wires might have """
+    if isinstance(wires, (str, Number)):
+        return (wires,)
 
     if isinstance(wires, Wires):
-        # if input is already a Wires object, just return its wire tuple
-        return wires.labels
+        return tuple(wires)
 
-    if isinstance(wires, (Number, str)):
-        # interpret as a single wire
-        return (wires,)
-
-    if hasattr(wires, "shape") and wires.shape == tuple():
-        # Scalar NumPy array
+    if getattr(wires, "shape", None) == tuple():
         return (wires.item(),)
 
     if isinstance(wires, Iterable):
-        # wires is an iterable, presumably of Wires, Number or scalar NumPy array instances
-        if all(isinstance(w, Wires) for w in wires):
-            # if the elements are themselves Wires objects, merge them to a new one
-            merged = tuple(w for wires_ in wires for w in wires_.tolist())
-        elif all(
-            isinstance(w, (str, Number)) or (getattr(w, "shape", None) == tuple()) for w in wires
-        ):
-            # if the elements are strings or numbers, turn iterable into tuple
-            merged = tuple([w.item() if isinstance(w, np.ndarray) else w for w in wires])
-        else:
-            merged = None
-
-        if merged is not None:
-            # check that all wires are unique
-            if len(set(merged)) != len(merged):
-                raise WireError("Wires must be unique; got {}.".format(merged))
-            return merged
+        out = tuple()
+        for w in wires:
+            out += _process(w)
+        if len(out) != len(set(out)):
+            raise WireError("Wires must be unique; got {}.".format(out))
+        return out
```

Comment:
Oh, I really like this recursive solution!

Reply:
Thank you :)